bench/stb/optimized/stb_connected_components.c.ll

```diff
@@ -657,11 +657,12 @@ for.body:                                         ; preds = %if.end39, %if.end12
   %indvars.iv75 = phi i64 [ %i.0, %if.end39 ], [ %indvars.iv.next76, %if.end124 ]
   %indvars.iv = phi i64 [ %j.0, %if.end39 ], [ %indvars.iv.next, %if.end124 ]
   %k.072 = phi i32 [ 0, %if.end39 ], [ %inc, %if.end124 ]
-  %7 = add nsw i64 %indvars.iv, %3
+  %7 = add nuw nsw i64 %indvars.iv, %3
+  %idxprom42 = and i64 %7, 4294967295
```

Comment:
Regression.

Reply:
I think to avoid this regression it might be necessary to represent the nneg flag in scev.

Reply:
Though it's very hard to understand what is actually going on from the end result -- my best guess is that we have something like zext (add i32) here originally and perform an sext IV widening and treat this as a truncating use, resulting in zext (trunc (add i64)). But I'm not confident this is the right guess.

Reply:
I think this demonstrates the issue:
```llvm
target datalayout = "n64"

define void @test(i32 %x, i32 %y, i32 %z) {
entry:
  br label %loop

loop:
  %iv = phi i32 [ 0, %entry ], [ %iv.inc, %loop ]
  %iv.sext = sext i32 %iv to i64
  call void @use(i64 %iv.sext)

  %add = add nuw nsw i32 %iv, %z
  %add.ext = zext nneg i32 %add to i64
  call void @use(i64 %add.ext)

  %iv.inc = add i32 %iv, 1
  %cmp = icmp ult i32 %iv, 100
  br i1 %cmp, label %loop, label %exit

exit:
 ret void
}

declare void @use(i64)
```

We perform an sext widening (apparently, we always prefer sext widening if there is at least one sext) and insert an unnecessary trunc after the add.

Reply:
This is most likely fixed by https://github.com/llvm/llvm-project/commit/c16559137cd1ad08ef5934558829d2ff5d227d70.